### ISSUE TITLE
slows down session restore dramatically

### ISSUE THREAD

**Issue by leedoyle** [rNeomy/tab-suspender#26]
When restoring the session with 150 tabs with SessionManager, Tab Suspender slows down restoring of the session dramatically, even though I've set it not to autoload the suspended tabs on restart and it doesn't unsuspend any of the suspended tabs. In fact, restoring the session with 150 unloaded tabs takes 3 minutes when normally it only takes 10-15 seconds at worst.

**Comment by tonsimple** [comment]
Can confirm.

 session restoring either through Session Manager extension or through native firefox means becomes sluggish as soon as tab count goes over a 100

Also noticed that if the profile was configured to start offline (firefox -P instrument or via offline switch extension) the loading times become unbearable and a "script is taking too long" error arises (I did not have the prudence to write the error down, alas)

My configuration is
Qubes, Debian 8, firefox ESR

tested on "naked", fresh install firefox restoring session using normal means

firefox + session manager

and 
firefox + session manager + offline switch

**Comment by rayman89** [comment]
Can confirm as well im on widows 7 x64 using firefox beta 55 x64 using session manager addon.

**Comment by tonsimple** [comment]
@rNeomy Also just moved to tabmixPlus and am observing precisely same behavior (clean install, then install tabmix plus, restart, then the tab suspender, restart firefox again, then create a lot (~200 tabs total) tabs and windows... and voila, next restore crawls to a halt)

**Comment by grahamperrin** [comment]
2017-07-14: 

> … beta 55 …

2017-07-22: [resolution and fix](https://bugzilla.mozilla.org/show_bug.cgi?id=906076#c256) under _906076 - (lazytabs) Virtual tabs - lazily create linkedBrowser and other dependent elements for tabbrowser tabs to improve startup performance_. Related: 

- [The New Firefox and Ridiculous Numbers of Tabs – Dietrich Ayala](https://metafluff.com/2017/07/21/i-am-a-tab-hoarder/) (2017-07-21)

So maybe review this issue with a more recent pre-release of 55; or with the release, which is scheduled for 2017-08-08.

**Comment by rNeomy** [comment]
From 0.1.9, there will be no restoring by this add-on, instead of the suspended tabs are listed in a new window. If you don't have an extension that restores the suspended tabs there should be no performance issue; https://addons.mozilla.org/en-US/firefox/addon/tab-suspender-tab-unloader/versions/beta

**Comment by rayman89** [comment]
@rNeomy Sorry but I didnt quite understand. What would happen when you unload a tab? and how do you restore the tab?

**Comment by tonsimple** [comment]
@rNeomy Sorry, I didn't "get" that either.
Will you have your own session management/restore, like Great Suspender in chrome, or will I have to restore suspended tabs separately from non-suspended ones?

I am very confused now.

**Comment by indeedwatson** [comment]
I experience a 15s slowdown when opening FF 57 with this addon installed. I'm not restoring session through session manager nor through the crashed session native dialogue, but I do have about 130 tabs.

**Comment by rayman89** [comment]
@indeedwatson If you start you firefox and it automatically opens 130 tabs then you are using the vanilla firefox sessoion restore there is no dialog if you set it to restore session in options>general>when firefox starts: restore windows and tabs the dialog appears after a crash only if im not mistaken.

**Comment by indeedwatson** [comment]
Yeah you're right @rayman89 Sadly that makes this addon pretty much unusable for now.

**Comment by sithdaddy** [comment]
In my case , whenever Firefox is not running AND i have a lot of tabs as part of a Firefox session  AND open a link from an email OR from skype  AND Firefox is called to open the that link , that is when it takes Firefox a long time to load .

When i start from the firefox icon and it restores the same session , no problem.

This addon seems to have some problems when firefox is started by an external program it seems.

I disabled all other addons to exclude the possibility of addons not working well with each other.

**Comment by irvinm** [comment]
There has to be a better way of handling this situation.  I have only started using this add-on and have only had 50-60 tabs open so far and have noticed no problems (FF57 Beta).  When FF57 goes live, I expect to move everything over and that will be 400+ tabs and having to deal with "suspended" vs. "non-suspended" tabs differently (which I expect to have 90%+ of my tabs suspended at any given moment) would be disastrous after each upgrade\patch of Firefox.

For the last Beta build pushed, I reloaded all of my tabs (Got them un-suspended) before restarting to avoid this situation but that is only going to work with a reasonable number of tabs and not the larger number of tabs I usually have.

**Comment by leedoyle** [comment]
@irvinm there is, it's called tabs.discard API and is landed in Firefox 58. I guess that should do it.

**Comment by irvinm** [comment]
Yeah, I know (https://github.com/rNeomy/tab-suspender/issues/31#issuecomment-340802535).  I tried using the native implementation with FF58b1 and it didn't work for me (https://github.com/rNeomy/tab-suspender/issues/61#issuecomment-343271138).  Seeing as discard is new to FF58, I will let it bake some more to see how things play out.

rNeomy has already said that the use of the native discard is in the code, so there is definitely hope!

**Comment by irvinm** [comment]
I played with FF58b and FF59a a little this morning and everything worked fine with either the "native" option enabled or disabled. However, I am unsure if there is any visual way to tell if the addon is using the native browser implementation of discard or not as the standard discarded page is loaded in both situations.

I didn't try it with my large browser session (500+ tabs), but merely tried a single tab. [Maybe the performance of the native implementation or the resource savings would be easier to see with the larger session]

**Comment by sebschrader** [comment]
I can confirm this issue both with Firefox 57 and Firefox 58 Beta (which contains the new `tab.suspend` API).

When I initially installed the add-on everything was fine, but upon browser restart everything went haywire. I have a browser session with 800+ tabs. CPU and memory usage always went up until the system was unusable, because Firefox was consuming most of the RAM. After disabling the add-on before performing session restore everything was fine again.

I installed this add-on precisely because I'm accumulating so many tabs over time, but this makes this add-on pretty much unusable.
  
EDIT: The authors new add-on [Auto Tab Discard](https://github.com/rNeomy/auto-tab-discard), seems to work fine however.
